texar/core/layers.py

```diff
@@ -664,6 +664,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
                 layer_output = layer(input)
                 layer_outputs.append(layer_output)
 
+        # the merge dimension cannot be determined until we get the output from
+        # individual layers.
+        # In case of reduce pooling operations, feature dim is removed and
+        # channel dim is merged.
+        # In non-reduce pooling operations, feature dim is merged.
+        if self._dim is None:
+            self._dim = layer_outputs[0].dim() - 1
```

Comment:
Prefer not to change `self._dim`. It might be possible for the user to pass in tensors of different ranks.

Reply:
I considered that possibility. `dim` arguments to torch functions in `forward` depends on the size of the input tensor. I am thinking to remove `self._dim` and make it function variable. In that case, we should be not be taking `dim` argument in the constructor. Also, `dim` is not being passed with any value during construction of `MergeLayer`. I think we can safely remove this. However, if you consider `texar-TF` interface compatibility, we may have to keep this.

Reply:
We don't really need to move the `dim` argument to `forward`. What I meant was:
```python
dim = self._dim is self._dim is not None else -1
```
and use `dim` in the merging ops.

Reply:
Done!